pkg/controller/common.go

```diff
@@ -746,6 +769,27 @@ func getRankedNodesForReplicaAttachments(ctx context.Context, azr azReconciler,
 	return nodes[:min(len(nodes), numReplica)], nil
 }
 
+func getSupportedZones(nodeSelectorTerms []v1.NodeSelectorTerm) map[string]bool {
+	// Get the list of supported zones for pv
+	supportedZones := map[string]bool{}
+	if len(nodeSelectorTerms) > 0 {
+		for _, term := range nodeSelectorTerms {
+			if len(term.MatchExpressions) > 0 {
+				for _, matchExpr := range term.MatchExpressions {
+					if matchExpr.Key == consts.AzureDiskCSITopologyKey {
```

Comment:
Should we check for `const.WellKnownTopologyKey` as well?

Reply:
Yeah, I think we should only check for consts.WellKnownTopologyKey, because no matter if it's SXS deployment or an isolated v2 deployment, this key will be applied to both the PVs and the nodes. What do you say?

Reply:
Hmmm. I guess that all depends on whether or not it is included in the PV. Looking at this more closely, the driver's implementation of `CreateVolume` returns topology using the driver-specific topology key. I just inspected a PV in my multi-zone cluster and see that only the driver-specific topology key is used in the node affinity term. It appears that the well-known key is not used with PVs, so we must only use the driver-specific key here.

Reply:
Yeah that makes sense. Although, I do see the WellKnownTopologyKey for the PVs in my multi zone cluster. Will check why that is so.

<img width="284" alt="image" src="https://user-images.githubusercontent.com/74526631/152242804-fc4bd760-1787-4322-8e2d-f0a30dfaf9b8.png">

Reply:
Made changes to 
1. Use topology.disk.csi.azure.com ortopology.disk2.csi.azure.com/zone as topology keys when filtering out the node selector terms from a pv
2. Use WellKnownTopologyKey to filter out the matching nodes. As per this code, the nodes on multi-az cluster should have the WellKnownTopologyKey  annotation : https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/fd82058e7b77c06234a64160a8d83965724218e3/pkg/azuredisk/nodeserver_v2.go#L399

Reply:
> 2\. Use WellKnownTopologyKey to filter out the matching nodes. As per this code, the nodes on multi-az cluster should have the WellKnownTopologyKey  annotation : https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/fd82058e7b77c06234a64160a8d83965724218e3/pkg/azuredisk/nodeserver_v2.go#L399

According to the [Kubernetes documentation](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone), the `topologykubernetes.io.zone` is populated by information from the `cloudprovider` and the CSI driver. While these values should match, we should clearly understand that the response from `NodeGetInfo` is not (necessarily) the source of the well-known node topology value.

According to the [NodeGetInfo documentation](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetinfo), these values are optional, however it does appear to apply the driver-specific topology label after installation. It's possible that `NodeGetInfo` value may override the value provided by the `cloudprovider`, but I'm not sure we should rely on that assumption. Since the well-known topology value returned by the driver is derived from the `cloudprovider` (namely, the `cloud-provider-azure` library), the two should never differ.

Reply:
> I do see the WellKnownTopologyKey for the PVs in my multi zone cluster. Will check why that is so.

What I thought was a multi-zone cluster turned out to be single-zone - my bad. In the single-zone case, it appears that the well-known topology key is empty.

Reply:
FYI- For a PV in a single zone cluster, only the driver-specific topology key is included and its value is empty:

```
Capacity:          10Gi
Node Affinity:
  Required Terms:
    Term 0:        topology.disk.csi.azure.com/zone in []
```

Reply:
The labels for a node in a single-zone cluster include the following:

```
topology.disk.csi.azure.com/zone=
topology.kubernetes.io/region=northeurope
topology.kubernetes.io/zone=2
```

Interestingly, the well-known label value is not empty, but also does not following the standard "<region>-<zone>: format. I wonder if this is a fault domain?

Reply:
Multi-zone node labels for reference:

```
topology.disk.csi.azure.com/zone=northeurope-1
topology.kubernetes.io/region=northeurope
topology.kubernetes.io/zone=northeurope-1
```

Reply:
Thanks for sharing the different annotations.

So, I guess since the WellKnownTopology  is set by the cloudprovider and the nodes will always be annotated by that,   it's safe to go with the this key for filtering out the nodes.

For single zone cluster, as you mentioned, the value of the topologyKey annotation for the pv would be empty, so the len(compatibleZones) would be 0. In that case, we are anyway not doing any zone based filtering, so that should work too.

Reply:
The cluster was built using `aks-engine`. It's `edreed-k8s-single-az` in the `edreed-k8s-single-az-rg` resource group.